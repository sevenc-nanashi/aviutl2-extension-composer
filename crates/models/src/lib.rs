mod generated {
    include!(env!("SCHEMA_RS_PATH"));

    impl Copy for Uint {}
    impl PartialEq for DataDirRelativePath {
        fn eq(&self, other: &Self) -> bool {
            self.0 == other.0
        }
    }
    impl Eq for DataDirRelativePath {}
}
pub use generated::*;
pub use overrides::*;

#[derive(Debug, Clone, Copy, PartialEq, Eq, strum::Display, strum::EnumString)]
pub enum RootType {
    #[strum(to_string = "$alias")]
    Alias,
    #[strum(serialize = "$default", to_string = "$palette")]
    Palette,
    #[strum(to_string = "$figure")]
    Figure,
    #[strum(to_string = "$language")]
    Language,
    #[strum(to_string = "$plugin")]
    Plugin,
    #[strum(to_string = "$script")]
    Script,
    #[strum(to_string = "$transition")]
    Transition,
    #[strum(to_string = "$theme")]
    Theme,
    #[strum(to_string = "$data")]
    Data,
}

mod overrides {
    use super::RootType;
    #[allow(non_camel_case_types)]
    pub type u64 = ::std::primitive::u64;

    pub type Url = ::url::Url;

    #[derive(Debug, Clone, PartialEq, Eq)]
    pub struct DataDirRelativePathInner {
        pub root: RootType,
        pub path: String,
    }
    impl DataDirRelativePathInner {
        pub fn is_directory(&self) -> bool {
            self.path.ends_with('/')
        }
    }
    impl serde::Serialize for DataDirRelativePathInner {
        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
        where
            S: serde::Serializer,
        {
            serializer.serialize_str(&format!("{}/{}", self.root, self.path))
        }
    }
    impl<'de> serde::Deserialize<'de> for DataDirRelativePathInner {
        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
        where
            D: serde::Deserializer<'de>,
        {
            let ss = String::deserialize(deserializer)?;
            let mut parts = ss.splitn(2, '/');
            let root_str = parts
                .next()
                .ok_or_else(|| <D::Error as serde::de::Error>::custom("missing root"))?;
            let path = parts
                .next()
                .ok_or_else(|| <D::Error as serde::de::Error>::custom("missing path"))?
                .to_string();
            if path.is_empty() {
                return Err(<D::Error as serde::de::Error>::custom("empty path"));
            }
            if path.contains('\\') {
                return Err(<D::Error as serde::de::Error>::custom(
                    "path contains backslash",
                ));
            }
            if path.contains("..") {
                return Err(<D::Error as serde::de::Error>::custom(
                    "path contains parent directory reference",
                ));
            }
            if path.starts_with('/') {
                return Err(<D::Error as serde::de::Error>::custom(
                    "path starts with a slash",
                ));
            }
            if path.contains("//") {
                return Err(<D::Error as serde::de::Error>::custom(
                    "path contains consecutive slashes",
                ));
            }

            let root = root_str
                .parse::<RootType>()
                .map_err(|_| <D::Error as serde::de::Error>::custom("invalid root"))?;
            Ok(DataDirRelativePathInner { root, path })
        }
    }

    pub mod license {
        macro_rules! create_license_struct {
            ($name:ident, $str_val:expr) => {
                #[derive(Debug, Clone, PartialEq, Eq)]
                pub struct $name;

                impl serde::Serialize for $name {
                    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
                    where
                        S: serde::Serializer,
                    {
                        serializer.serialize_str($str_val)
                    }
                }

                impl<'de> serde::Deserialize<'de> for $name {
                    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
                    where
                        D: serde::Deserializer<'de>,
                    {
                        let ss = String::deserialize(deserializer)?;
                        if ss == $str_val {
                            Ok($name)
                        } else {
                            Err(<D::Error as serde::de::Error>::invalid_value(
                                serde::de::Unexpected::Str(&ss),
                                &concat!("`", $str_val, "`"),
                            ))
                        }
                    }
                }
            };
        }
        create_license_struct!(Free, "free");
        create_license_struct!(Nicovideo, "nicovideo");
        create_license_struct!(Custom, "custom");
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parsing_example_manifest() {
        insta::assert_debug_snapshot!(
            serde_yml::from_str::<Manifest>(include_str!("../../../docs/examples/manifest.yml"))
                .unwrap()
        );
    }

    #[test]
    fn test_parsing_example_registry() {
        insta::assert_debug_snapshot!(
            serde_yml::from_str::<Registry>(include_str!("../../../docs/examples/registry.yml"))
                .unwrap()
        );
    }
}
