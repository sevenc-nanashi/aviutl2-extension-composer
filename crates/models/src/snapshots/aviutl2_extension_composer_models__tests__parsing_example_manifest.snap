---
source: crates/models/src/lib.rs
expression: "serde_yml::from_str::<Manifest>(include_str!(\"../../../docs/examples/manifest.yml\")).unwrap()"
---
Manifest {
    authors: [
        ManifestAuthorsItem {
            name: Single(
                "Nanashi.",
            ),
            url: Some(
                HttpUrl(
                    Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "sevenc7c.com",
                            ),
                        ),
                        port: None,
                        path: "/",
                        query: None,
                        fragment: None,
                    },
                ),
            ),
        },
    ],
    bundles: None,
    changelogs: [],
    configurations: [],
    description: Some(
        Single(
            "Rustで書かれたFFmpeg出力プラグインです。\n...\n（省略）\n...",
        ),
    ),
    disposables: [],
    homepage: Some(
        "https://github.com/sevenc-nanashi/aviutl2-rs",
    ),
    id: ManifestId(
        "sevenc_nanashi-aviutl2_rs-ffmpeg_output",
    ),
    license: Nicovideo {
        id: "sm45355531",
        name: Nicovideo,
        omittable: Some(
            true,
        ),
        text: None,
    },
    manifest_url: Some(
        HttpUrl(
            Url {
                scheme: "https",
                cannot_be_a_base: false,
                username: "",
                password: None,
                host: Some(
                    Domain(
                        "raw.githubusercontent.com",
                    ),
                ),
                port: None,
                path: "/sevenc-nanashi/aviutl2-extension-composer/refs/heads/main/docs/examples/manifest.yml",
                query: None,
                fragment: None,
            },
        ),
    ),
    manifest_version: 1,
    name: "Rusty FFmpeg Output",
    previous_versions: [],
    resources: [
        ManifestResourcesItem {
            destination: DataDirRelativePath(
                DataDirRelativePathInner {
                    root: Plugin,
                    path: "rusty_ffmpeg_output.auo2",
                },
            ),
            scope: None,
            sha256: None,
            source: "https://sevenc7c.example.com/rusty_ffmpeg_output/rusty_ffmpeg_output.auo2",
        },
    ],
    summary: Single(
        "Rustで書かれたFFmpeg出力プラグイン",
    ),
    tags: [
        Single(
            "#plugin-output",
        ),
    ],
    version: Version(
        "0.1.0",
    ),
    version_number: Some(
        Uint(
            1,
        ),
    ),
}
